src/mlpack/methods/ann/dists/bernoulli_distribution_impl.hpp

```diff
@@ -51,8 +51,8 @@ DataType BernoulliDistribution<DataType>::Sample() const
 {
   DataType sample = arma::randu<DataType>
       (probability.n_rows, probability.n_cols);
-
-  for (size_t i = 0; i < sample.n_elem; i++)
+  #pragma omp parallel for
+  for (omp_size_t i = 0; i < sample.n_elem; i++)
```

Comment:
I think it would depend on the number of samples, right? If there are maybe only 10 samples, there may not be a difference in performance. However, as the number increases & the overload of creating the threads becomes less & less relevant, we would see a greater increase in performance

Reply:
You didn't take into account that RAM frequency is much lower than CPU frequency and the loop requires 2 new values each iteration. Actually, there are a great number of factors such as memory frequency, memory bandwidth, CPU frequency, CPU cache size and so on.

I wrote a simple test which measures the duration of a similar loop.

https://gist.github.com/lozhnikov/3486432717ea04f25a722c97fbd79edd

(Weird, I couldn't upload the file, so I created a gist)

Here are the results:

I used a system with core-i7 2600K and 16GB DDR3 memory.

Look at the "Parallel (s)" and "Sequential (s)" columns.

### g++ 9.2.1
```
g++ -O3 -fopenmp speedup_test.cpp -lgomp -o speedup_test_g++
./speedup_test_g++ 
        Size   Count Par. Parallel (s)   Count Seq. Sequential (s)
          10            4 3.410000e-07            3 7.200000e-08
         100           51 2.920000e-07           51 1.190000e-07
        1000          497 1.979000e-06          551 8.550000e-07
       10000         4999 1.912400e-05         4992 8.455000e-06
      100000        50072 1.392650e-04        50460 8.161700e-05
     1000000       500442 1.370827e-03       500021 1.592891e-03
    10000000      4998118 1.238903e-02      5001644 1.242419e-02
   100000000     50002335 1.249063e-01     49990234 1.196964e-01
```

### clang 9.0.1
```
clang++ -O3 -fopenmp speedup_test.cpp -lgomp -o speedup_test_clang++
        Size   Count Par. Parallel (s)   Count Seq. Sequential (s)
          10            4 5.200000e-07            6 3.320000e-07
         100           43 1.970000e-07           52 7.500000e-08
        1000          509 7.580000e-07          536 2.570000e-07
       10000         5015 9.891000e-06         5062 3.591000e-06
      100000        49738 7.213000e-05        50136 4.434200e-05
     1000000       499560 1.313246e-03       499601 1.290657e-03
    10000000      4998336 1.029061e-02      5001621 1.008930e-02
   100000000     49997002 9.726773e-02     50000397 9.817899e-02
```

There's hardly any difference.

Reply:
I deleted the last comment due to wrong values.

Reply:
**Added another comment. Now the test seems correct.**

You didn't take into account that RAM frequency is much lower than CPU frequency and the loop requires 2 new values each iteration. Actually, there are a great number of factors such as memory frequency, memory bandwidth, CPU frequency, CPU cache size and so on.

I wrote a simple test which measures the duration of a similar loop.

https://gist.github.com/lozhnikov/3486432717ea04f25a722c97fbd79edd

(Weird, I couldn't upload the file, so I created a gist)

Here are the results:

I used a system with core-i7 2600K and 16GB DDR3 memory.

Look at the "Parallel (s)" and "Sequential (s)" columns.

### g++ 9.2.1
```
g++ -O3 -fopenmp speedup_test.cpp -lgomp -o speedup_test_g++
./speedup_test_g++ 
        Size   Count Par. Parallel (s)   Count Seq. Sequential (s)
          10            6 1.880000e-07            4 5.900000e-08
         100           49 1.600000e-07           49 1.070000e-07
        1000          502 1.101000e-06          500 8.470000e-07
       10000         4964 1.166800e-05         4968 8.230000e-06
      100000        49841 2.085290e-04        49958 8.180200e-05
     1000000       500518 1.397976e-03       499577 1.487266e-03
    10000000      4997717 1.245767e-02      4998914 1.237367e-02
   100000000     49997606 1.240536e-01     49997601 1.196321e-01
```

### clang 9.0.1
```
clang++ -O3 -fopenmp speedup_test.cpp -lgomp -o speedup_test_clang++
        Size   Count Par. Parallel (s)   Count Seq. Sequential (s)
          10            4 4.780000e-07            5 1.270000e-07
         100           47 2.260000e-07           51 7.800000e-08
        1000          520 1.019000e-06          489 2.660000e-07
       10000         4965 1.011800e-05         4993 3.656000e-06
      100000        49904 6.398400e-05        49987 4.442300e-05
     1000000       499302 1.313890e-03       498852 1.307853e-03
    10000000      4999504 1.020465e-02      4998569 1.076408e-02
   100000000     50005009 9.804719e-02     49992123 9.778105e-02
```

There's hardly any difference.

Reply:
You're right. I just ran a similar test on my machine:
1.6GHz dual-core Core i5, 4GB DDR3 mem

I underestimated the amount of data you would need to see a significant difference. You only see a real difference when the number of items is 10M+

If the comparison is really quick, then I can remove the parallel part, and reduce overhead. What do you say?

Reply:
Yeah, I think you can remove the parallel part here.